### DISCUSSION TITLE
Livewire is getting corroupted after a refresh. All documents attached.

### DISCUSSION THREAD

**Discussion by debjit** [livewire/livewire#3468] [1. Help]
## This is my first attempt working with livewire. 
This is a quiz so I don't know much about livewire. Please help.
So The scenario is when people give a quiz without page refresh it is ok. But when you refresh the error come up after submitting one or two response saying data has been corrupted.
Here is what I am using in the code.
[controller view and migration](https://gist.github.com/debjit/5e06ecc6cd1c2d7761c4ccbdb1177b48)
Here is the error msg from the log file.
[Error ](https://gist.github.com/debjit/cb866f14cdf67f6144f884236a3be0dd)

Please let me know if you want any other information. Thank you

**Comment by Qirel** [comment]
There's a lot going on here - which method, specifically, gets called when you run into that issue? It's a tad hard to determine your exact issue without at least knowing which method gets called. 

Here's a little explanation I wrote up as to why the error in general can get thrown, with a general example. It might give you a clue on where to begin looking. 

---------------------

The "Livewire encountered corrupt data when trying to hydrate the ... component" can be a bit tricky to debug at times, but its occurring when Livewire re-hydrates, and it cannot match the checksum of the passed data, with the checksum that was originally sent. These can only match when the data in the request was not tampered with (its a security measure). But its not always that you *actually* tampered with the data, and still the checksums isn't matching.

For example, have a look at this simple test component, that just has a collection of two items in it. To visualize that we actually re-render (to prove if the refresh works), we keep track of the count with a property `$renderCount` that is just refreshed on each iteration.
```php
use Illuminate\Support\Collection;
use Livewire\Component;

class Test extends Component
{
    public Collection $test;
    public int $renderCount = 0;

    public function mount()
    {
        $this->test = collect([
            'foo' => 'bar',
            1 => 'baz',
        ]);
    }

    public function render()
    {
        $this->renderCount++;
        return view('livewire.test.test');
    }
}
```
And its corresponding view,
```html
<div>
    <p>We have rendered {{ $renderCount }} times</p>
    <button wire:click="$refresh">Refresh me!</button>
    @dump($test)
</div>

```
It's all peaches on the initial load. The dump is what one would expect, the collection of the array. However, when you click the button, the daunty exception is thrown,

> ## Livewire\Exceptions\CorruptComponentPayloadException  
> Livewire encountered corrupt data when trying to hydrate the [test] component. Ensure that the [name, id, data] of the Livewire component wasn't tampered with between requests.

And here's the kicker - the test-component we're using is not using `$data`, `$id` or `$name` as one of its properties like the exception suggests - so what's up with that? 

Well, the short answer is that JavaScript is the sinner on this one. It has to do with how JavaScript handles objects (specifically the order of items). If we go back to our class, and replace the collection with inversed order, making the numbers come first, 
```php
$this->test = collect([
    1 => 'baz',
    'foo' => 'bar',
]);
```
..and repeat the process of loading up and refreshing, we'll see that the initial load is as before - as we expect. But when we click the button to refresh, the counter increments and we no longer get an error - it suddenly just works, ***why is that?***

### Inspecting the data

Let's have a look. I changed the order of the collection back so we can reproduce the error. Here's an image from the Laravel Debugbar after the first, initial load.

![image](https://user-images.githubusercontent.com/16939228/128152719-26cf93fd-4200-4fb3-9466-4d393fe37722.png)

Look at the data of the collection - its matching what we have in the component. Now open up the devtools in your browser, and find the network tab. Hit the "Refresh me" button on our page to trigger the refresh of the component, and as expected the exception is thrown. In your network-tab, you now have a request with a status-code of 500 that is the request to the component. Let's study the request payload (in Chrome, click on the request, open the "Headers" tab and scroll to the bottom to see "Request Payload"). 

![image](https://user-images.githubusercontent.com/16939228/128152637-d57b5e95-c17b-4c22-a5d2-c49b1369556e.png)

We're looking for the `data` property under `serverMemo`.  There we'll see our property `test` -- but its now in the reverse order! This caused by JavaScript, when you have an object with keys that are mixed numbers of strings. See the image from my console here how it changed order.

![image](https://user-images.githubusercontent.com/16939228/128147001-5b4c7fc9-0016-498c-be77-c87dac46716a.png)

#### Why did this happen?

When Livewire first renders, it passes all the data in the public properties to JavaScript - so that we keep track of the current state of the component. JavaScript will then internally re-order our `$test` collection, numbers first, then strings - and when we click the button to refresh, JavaScript is passing the data back to the server along with the original checksum. In the backend, Livewire will accept the request, obtain the checksum of the data that was the previous state, and match it up with the checksum you sent from the last request. But, because the order changed - the checksum is now *different* between the two - even if it is the exact same data (just re-ordered), and because the checksums do not match, the exception is thrown.

The checksum can be different for any number of reasons - but common for them all is that the data that was in the previous state was changed between the two requests. In this example, I already knew the error would be thrown because of the re-ordering of items in JavaScript, but finding these differences can be tough if you have bigger and more complex components.

## How to debug this?

The checksum is generated in the [ComponentChecksumManager class](https://github.com/livewire/livewire/blob/master/src/ComponentChecksumManager.php), so that's where we can start looking. The easiest way to debug this, is by logging down the `$memoSansChildren` variable for each time it generates a checksum. It can be a good idea to limit the logging to the specific component we're having trouble with to not flood our logs. My component is called `Test\Test`, so its name is `test.test`. So before the `return` inside `ComponentChecksumManager::generate()` , I can put this little statement,

```php
if ($fingerprint['name'] === 'test.test')
    \Log::debug(json_encode($memoSansChildren));
```

Then load up my component with the error, and refresh it (or perform the action that throws this exception). When the exception is thrown, I can inspect my logs to compare the two memos. In my case, they looked like this,

```json
{"errors":[],"htmlHash":"9fab4b3b","data":{"test":{"foo":"bar","1":"baz"},"renderCount":1},"dataMeta":{"collections":["test"]}}  
{"errors":[],"htmlHash":"9fab4b3b","data":{"test":{"1":"baz","foo":"bar"},"renderCount":1},"dataMeta":{"collections":["test"]}}  
```

As we can see, the two `data.test` properties hold the same data, but they are in different order. 

If you have larger components with a lot of data, you can run it through a text-comparing tool like WinMerge to spot the difference - the two JSONs should not be different (if they are equal, there are no exceptions). When comparing the two JSON strings, it can be helpful to pretty-print them first, so spotting different lines becomes easier. 

So now we have figured out what the problem was in this test-component, and as we touched on before, the solution here is to ensure the order of the data in the collection isn't internally changed by JavaScript (means we put the data in PHP in the same order as JavaScript would sort it). 

Once you are done and have fixed the error, remember to go back and reset the `ComponentChecksumManager::generate()` method to its original state, we shouldn't be changing things directly in the framework like that - and it will keep logging until you remove it..  😃 

Hope this gave some insight on why this "obscure" exception is thrown, and that you now have a tool to debug it without much guesswork.
